### PR TITLE
persist: add a new source operator for mirroring a persisted stream

### PR DESCRIPTION
**Pull request by danhhz**

### CODE REVIEW

File: src/persist/src/operators/source.rs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+use crate::indexed::TailEvent;
+use crate::{operators, Token};
+
+/// A Timely Dataflow operator that mirrors a persisted stream.
+pub trait PersistedSource<G: Scope<Timestamp = u64>, K: Data, V: Data> {
+    /// Emits a snapshot of the persisted stream taken as of this call and tails
+    /// any new data added to the persisted stream after that.
+    fn persisted_source(
+        &mut self,
+        read: Token<StreamWriteHandle<K, V>, StreamReadHandle<K, V>>,
```
Comment (ruchirK) [inline src/persist/src/operators/source.rs:29]: nit: I think this should be named `token` to match whats below?

File: src/persist/src/operators/source.rs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+                            TailEvent::Records(mut records) => {
+                                session.give_vec(&mut records);
+                                activator.activate();
+                            }
+                            TailEvent::Sealed(ts) => {
+                                cap.downgrade(&ts);
+                            }
+                        },
+                        Err(TryRecvError::Empty) => {
+                            activator.activate();
```
Comment (ruchirK) [inline src/persist/src/operators/source.rs:84]: maybe we want to activate_after a few seconds here?

File: src/persist/src/operators/source.rs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+            let activator = self.activator_for(&info.address[..]);
+            let mut cap = Some(capability);
+            move |output| {
+                let mut done = false;
+                if let Some(cap) = cap.as_mut() {
+                    let mut session = output.session(cap);
+                    match tail_rx.try_recv() {
+                        Ok(e) => match e {
+                            TailEvent::Records(mut records) => {
+                                session.give_vec(&mut records);
```
Comment (ruchirK) [inline src/persist/src/operators/source.rs:76]: it took me a while to reason about what happens with multiple workers but I think:

each worker will register its own individual tail with their own mpsc queue and then any sent data will get broadcast over all of those queues to all of the workers, and we will end up with N copies of the data iff there are N workers is that right? this seems we want to only do all of this on worker 0 or something like that for the v1?

**Review by ruchirK — APPROVED** [review]
LGTM! Left one question about multi-worker sources. I had one other small, non-actionable thought: I'm a little hesitant about the word tail because Materialize already uses it in two other contexts - the tail sql command and file sources "with tail" and that's confusing as is.

I'm not sure what a better name would be though.

File: src/persist/src/operators/source.rs (hunk 2 above)
Comment (danhhz) [reply src/persist/src/operators/source.rs:84]: Done.

File: src/persist/src/operators/source.rs (hunk 3 above)
Comment (danhhz) [reply src/persist/src/operators/source.rs:76]: Done. Added an ignored test. There's a decent bit of work to do to fix it up, but I'd prefer to do it in a followup, if that's alright.

**Review by danhhz — COMMENTED** [review]
TFTR! As discussed offline, renamed tail to listen. Rebased on top of my other PR (one interesting conflict but resolution was straightforward) and then addressed review comments in a separate commit if you'd like to take a last look.

**Pull request merged**